draft-ietf-privacypass-integration.md

```diff
@@ -0,0 +1,224 @@
+   uint8 version;
+   uint16 suite;
+   uint8 public_key[Ne];
+} KeyConfig;
+~~~
+
+KeyConfig.version corresponds to the version of the key configuration
+that is being used, this parameter is left generic and may be altered if
+extensions to the protocol or server configurations are made.
+KeyConfig.suite corresponds to a POPRF ciphersuite from
+{{I-D.irtf-cfrg-voprf, Section 4}}, and KeyConfig.public_key corresponds
+to a serialized public key of length `Ne` bytes (denoted as a
+`SerializedElement` in {{I-D.irtf-cfrg-voprf, Section 2}}). In
+
+In order for higher-level applications to indicate which key
+configuration is being used, a common identifier, such as
+`id=SHA256(KeyConfig)`, should be used. Note that the ciphersuite that
+is used is determined entirely by the choice of `suite` in the server
+key configuration.
+
+## Issuance phase
+
+Let `info` be the agreed upon metadata between client and server, and
+let `config` be the server's chosen key configuration.
+
+First, a client configures its verifiable context using `config`:
+
+~~~
+client_context = SetupVerifiableClient(
+                  config.suite, config.public_key
+                 )
+~~~
+
+Likewise, the server creates its own context using `config` and the
+corresponding private key `key`:
+
+~~~
+server_context = SetupVerifiableServer(
+                  config.suite, key, config.public_key
+                 )
+~~~
+
+The client then creates an issuance request for a random value `nonce`
+as follows:
+
+~~~
+nonce = random(32)
+blind, blindedElement = context.Blind(nonce)
+~~~
+
+The client then sends `blindedElement` to the server. The server, upon
+receipt, evaluates the request:
+
+~~~
+evaluatedElement, proof = server_context.Evaluate(
+                           key, config.public_key,
+                           blindedElement, info
+                          )
+~~~
+
+The server sends both `evaluatedElement` and `proof` to the client.
+These are concatenated together. As the length of both is fixed, there
+is no ambiguity in parsing the result.
+
+The client then completes issuance as follows:
+
+~~~
+output = client_context.Finalize(
+          nonce, blind, evaluatedElement, info
+         )
+~~~
+
+This procedure may fail with an error (`VerifyError` or
+`DeserializeError`), in which case the issuance is said to have failed.
+The output of the issuance protocol is the concatenation of `nonce` and
+`output`, denoted as `token`:
+
+~~~
+struct {
+   uint8 nonce[32];
+   uint8 output[Nh];
+} Token;
+~~~
+
+where `Nh` is as defined in {{I-D.irtf-cfrg-voprf}}.
+
+## Redemption phase
+
+The client sends the `Token` to the server to verify locally. In
+particular, the server verifies the `Token` as follows:
+
+~~~
+valid = server_context.VerifyFinalize(
+         key, token.nonce, token.output, info
+        )
+~~~
+
+Redemption is considered successful if `valid` is true.
+
+# IANA Considerations
+
+This specification registers new ciphersuites for the Privacy Pass
```

Comment:
In general the form is: "This specification creates a new page named X in the Y category containing a registry named Z, with the following fields:".

In the case, maybe X = Privacy Pass, Y = Privacy Pass, Z = "Cipher Suites".

Also think about whether your cipher suites need a numeric identifier, or whether it is sufficient to give them strings.

Reply:
Why do we need these ciphersuites at all? Aren't they redundant with the POPRF suite?

Reply:
I think we do need a new IANA registry.  Without a registry, we would have to Update one or more RFCs every time we want to add a ciphersuite, which is laborious to execute and confusing to read.

Where that registry should be created depends on its intended contents.  If all future VOPRF suites as conceived in the VOPRF draft will continue to be exactly what you want for Privacy Pass, then the VOPRF draft can open the registry.  Otherwise (e.g. if we envision extending Privacy Pass to publicly verifiable ciphersuites), we'll need to define it here.

At the moment, the latter seems more likely.

Reply:
Yeah, for example in a future world where we have a Blind BLS suite and a POPRF suite, it seems like we'll want some containing ciphersuite that refers to one or the other.

Reply:
Sorry, I don't really follow. Why would those things be in the same registry? This document is really just a thin wrapper around the POPRF, and no updates would be needed if new POPRF suites were added. In other words, the POPRF document owns the registry for POPRF suites.

Reply:
Perhaps PMBTokens (an alternative to VOPRF that also matches the synchronous model) is a clearer example than Blind Signatures (which can also be made to fit into the synchronous model, but perhaps less elegantly).

Reply:
I see your point, but it's not clear to me we actually want a single registry to list different "modes" of privacy pass, e.g., one without private metadata (POPRF) and one with private metadata (PMBTokens). Why not just make that a separate thing altogether? What's the value of merging these things?

Reply:
At some layer we need to indicate what protocol is being used for a web/HTTP API. Be it PrivacyPassVOPRFP521 or PrivacyPassCoolPostQuantum or PrivacyPassBLS. Seems like something that the PrivacyPass WG should define somewhere as things that higher level things point to, though maybe we don't have that in the Integration doc and put it somewhere else?

Reply:
The goal mentioned in the proposed schedule is to support ecosystems containing a mixture of clients and issuers that support POPRF and non-POPRF ciphersuites.  That means one way or another, different ciphersuite families need to be accessible from the same server identity.

For clarity, it might be more helpful to frame this question in terms of the specific structs in the HTTP API draft.

Reply:
> The goal mentioned in the proposed schedule is to support ecosystems containing a mixture of clients and issuers that support POPRF and non-POPRF ciphersuites. That means one way or another, different ciphersuite families need to be accessible from the same server identity.

Sure, but they could so by using separate types of key configurations or otherwise making this decision at the application layer. The goal of the schedule doesn't seem to require that a joint registry be created.

> For clarity, it might be more helpful to frame this question in terms of the specific structs in the HTTP API draft.

I'm not sure. I expect many users of Privacy Pass to just use the wire formats described in this doc.

Reply:
> Be it PrivacyPassVOPRFP521 or PrivacyPassCoolPostQuantum or PrivacyPassBLS. Seems like something that the PrivacyPass WG should define somewhere as things that higher level things point to

@dvorak42 I guess what gives me pause is that putting everything behind a single registry implies that all values in that registry have similar properties, but that's not the case here. In contrast to, say, TLS algorithm registries, which are really just different algorithms with the same set of features. The things we're considering merging here have fundamentally different features (or at least they do in my opinion), so it's not clear to me that merging them is appropriate. (Sorry if I'm not making my thinking clear!)

Reply:
@chris-wood Maybe you could send a PR that updates draft-ietf-privacypass-http-api in the way you're thinking?